csharp/src/Ice/CollocatedRequestHandler.cs

```diff
@@ -20,143 +22,90 @@ internal class CollocatedRequestHandler : IRequestHandler
             _requestId = 0;
         }
 
-        public IRequestHandler? Update(IRequestHandler previousHandler, IRequestHandler? newHandler) =>
-            previousHandler == this ? newHandler : this;
-
-        public void SendAsyncRequest(ProxyOutgoingAsyncBase outAsync) => outAsync.InvokeCollocated(this);
-
-        public void AsyncRequestCanceled(OutgoingAsyncBase outAsync, Exception ex)
-        {
-            lock (this)
-            {
-                if (_sendAsyncRequests.TryGetValue(outAsync, out int requestId))
-                {
-                    if (requestId > 0)
-                    {
-                        _asyncRequests.Remove(requestId);
-                    }
-                    _sendAsyncRequests.Remove(outAsync);
-                    if (outAsync.Exception(ex))
-                    {
-                        Task.Run(outAsync.InvokeException);
-                    }
-                    _adapter.DecDirectCount(); // invokeAll won't be called, decrease the direct count.
-                    return;
-                }
-                if (outAsync is OutgoingAsync o)
-                {
-                    Debug.Assert(o != null);
-                    foreach (KeyValuePair<int, OutgoingAsyncBase> e in _asyncRequests)
-                    {
-                        if (e.Value == o)
-                        {
-                            _asyncRequests.Remove(e.Key);
-                            if (outAsync.Exception(ex))
-                            {
-                                Task.Run(outAsync.InvokeException);
-                            }
-                            return;
-                        }
-                    }
-                }
-            }
-        }
-
         public Connection? GetConnection() => null;
 
-        public void InvokeAsyncRequest(ProxyOutgoingAsyncBase outAsync, bool synchronous)
+        public ValueTask<Task<IncomingResponseFrame>?> SendRequestAsync(OutgoingRequestFrame outgoingRequestFrame,
+            bool oneway, bool synchronous, IInvocationObserver? observer)
         {
             //
             // Increase the direct count to prevent the thread pool from being destroyed before
             // invokeAll is called. This will also throw if the object adapter has been deactivated.
             //
             _adapter.IncDirectCount();
 
+            IChildInvocationObserver? childObserver = null;
             int requestId = 0;
-            try
+            lock (this)
```

Comment:
This isn't really needed since CollocatedRequestedHandler is internal and our code shouldn't lock it outside its impl. but perhaps we forbid this even for internal objects?

Reply:
I think we should eventually use the other patter everywhere unless there is a reason not to, but not need to fix in this PR either.

Reply:
Ok I'll fix but I don't agree 😄. I think it's ok for internal classes and in fact the implementation of the the .NET libraries use this in a number of classes as well... it's useful to reduce garbage IMO.